stdlib/public/core/HashedCollections.swift.gyb

```diff
@@ -561,13 +561,24 @@ public struct Set<Element : Hashable> :
   public func intersect<
     S : SequenceType where S.Generator.Element == Element
   >(sequence: S) -> Set<Element> {
-    let other = sequence as? Set<Element> ?? Set(sequence)
-    var newSet = Set<Element>()
-    for member in self {
-      if other.contains(member) {
-        newSet.insert(member)
+
+    // Attempt to iterate the smaller between `self` and `sequence`.
+    // If sequence is not a set, iterate over it because it may be single-pass.
+    var newSet: Set<Element>
+
+    if let other = sequence as? Set<Element> {
+      newSet = Set<Element>(minimumCapacity: min(other.count, count))
```

Comment:
@gribozavr the resulting size is between 0 and `min(other.count, count)`. 

The closer we allocate towards `min(other.count, count)`, the more we risk wasting space.
The closer we allocate towards 0, the more we risk wasting time resizing the storage.

I don't think there's a "right" way to do this. Before I make the change you suggested, just want to layout this trade off.